### PR TITLE
Asynchronous checkpointing

### PR DESCRIPTION
**Pull request by dhananjay-ram**
*Issue #, if available:*

*Description of changes:*
- wrote a new feature to enable deepspeed checkpoint saving asynchronously in the background during training, so the model training can continue while checkpoint is being saved. It can be enabled with "async_cpkt=true" in deepspeed config.
- Enable Huggingface style model saving only at the final step to unblock the GPU for training.
- added a utility script to save Huggingface style model from deepspeed checkpoint. 

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: training/docker/patches/async_torch_checkpoint_engine.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import torch
+import multiprocessing as mp
+from concurrent.futures import ProcessPoolExecutor
+from deepspeed.runtime.checkpoint_engine.torch_checkpoint_engine import TorchCheckpointEngine
+
+# Persistent background worker pool
+_DS_ASYNC_POOL = ProcessPoolExecutor(max_workers=1, mp_context=mp.get_context('spawn'))
```
Comment (bbowman223) [inline training/docker/patches/async_torch_checkpoint_engine.py]: ThreadPoolExecutor might be faster than ProcessPoolExecutor.  ProcessPoolExecutor requires pickling / serialization, whereas if using ThreadPoolExecutor the threads can share memory. You can look at https://github.com/pytorch/pytorch/blob/main/torch/distributed/checkpoint/state_dict_saver.py as a reference where they have both thread based and process based async executors.

Comment (dhananjay-ram) [reply training/docker/patches/async_torch_checkpoint_engine.py]: I used it to avoid GIL contention that can actually slower the whole process. Also I wanted to avoid any accidental corruption of checkpoint during GPU -> CPU transfer.

File: training/docker/patches/async_torch_checkpoint_engine.py
Comment (elvisnunez7) [inline training/docker/patches/async_torch_checkpoint_engine.py:1]: Let's only have patch files in `training/docker/patches/`, so let's create the folder `training/docker/plugins` and move this file there.

Comment (elvisnunez7) [inline training/docker/patches/async_torch_checkpoint_engine.py:1]: Let's add a module-level docstring explaining what this file does/is for and type annotations/return types to function signatures.

File: training/docker/patches/async_torch_checkpoint_engine.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+def _background_disk_write(state_dict, path):
+    """The heavy serialization task executed by the background process."""
+    try:
+        torch.save(state_dict, path)
+        print(f"[Async IO Worker] Successfully saved: {path}")
+    except Exception as e:
+        print(f"[Async IO Worker] Error saving {path}: {e}")
```
Comment (elvisnunez7) [inline training/docker/patches/async_torch_checkpoint_engine.py]: if `torch.save` fails here, training will continue without saving a checkpoint, and the error message is likely to get buried in the logs. Should we instead use the `Future` returned by `_DS_ASYNC_POOL.submit()` to check the save status before the next checkpoint, and kill training if a checkpoint didn't save? Otherwise, if we don't kill training due to a failed save, we should have a way to let the user know about the failure.

File: training/src/hmf/train/pt/trainer.py [hunk 3]
```diff
@@ -193,3 +193,18 @@ def compute_loss(self, model, inputs, *args, **kwargs):
...
 
         return loss
+
+    @override
+    def _save(self, output_dir: Optional[str] = None, state_dict=None):
+        # 1. Identify if this is the final save of the training run
+        # max_steps is the total number of steps defined in training args
+        is_final_step = self.state.global_step >= self.args.max_steps
+        if is_final_step:
+            print(f"\n[Final Save] Step {self.state.global_step} reached. Saving full HuggingFace-style model...")
```
Comment (elvisnunez7) [inline training/src/hmf/train/pt/trainer.py]: Use logger instead of print (see `training/src/hmf/train/hkd/trainer.py` for reference).

File: training/src/hmf/train/pt/trainer.py [hunk 4]
```diff
@@ -193,3 +193,18 @@ def compute_loss(self, model, inputs, *args, **kwargs):
...
+            # Call super()._save() which performs the standard HF saving logic
+            # (saving config.json, tokenizer, and the consolidated/sharded weights)
+            super()._save(output_dir, state_dict)
+        else:
+            print("skipping HF style saving ...")
```
Comment (elvisnunez7) [inline training/src/hmf/train/pt/trainer.py]: As above, use logger.

File: training/src/hmf/train/pt/trainer.py [hunk 5]
```diff
@@ -193,3 +193,18 @@ def compute_loss(self, model, inputs, *args, **kwargs):
             return loss / self.args.gradient_accumulation_steps
 
         return loss
+
+    @override
+    def _save(self, output_dir: Optional[str] = None, state_dict=None):
+        # 1. Identify if this is the final save of the training run
+        # max_steps is the total number of steps defined in training args
+        is_final_step = self.state.global_step >= self.args.max_steps
```
Comment (elvisnunez7) [inline training/src/hmf/train/pt/trainer.py]: `max_steps` may not always be defined. For example, users can instead specify the number of epochs to train for, so we should also handle this.

File: training/src/hmf/train/pt/trainer.py [hunk 6]
```diff
@@ -193,3 +193,18 @@ def compute_loss(self, model, inputs, *args, **kwargs):
             return loss / self.args.gradient_accumulation_steps
 
         return loss
+
+    @override
+    def _save(self, output_dir: Optional[str] = None, state_dict=None):
```
Comment (elvisnunez7) [inline training/src/hmf/train/pt/trainer.py]: Add docstring

File: training/src/hmf/train/pt/trainer.py [hunk 7]
```diff
@@ -193,3 +193,18 @@ def compute_loss(self, model, inputs, *args, **kwargs):
...
+        is_final_step = self.state.global_step >= self.args.max_steps
+        if is_final_step:
+            print(f"\n[Final Save] Step {self.state.global_step} reached. Saving full HuggingFace-style model...")
+            # Call super()._save() which performs the standard HF saving logic
+            # (saving config.json, tokenizer, and the consolidated/sharded weights)
+            super()._save(output_dir, state_dict)
+        else:
+            print("skipping HF style saving ...")
+
+        return
```
Comment (elvisnunez7) [inline training/src/hmf/train/pt/trainer.py]: `return` here is redundant; can remove it

File: training/scripts/convert_ds_to_hf.py
Comment (elvisnunez7) [inline training/scripts/convert_ds_to_hf.py:1]: Let's add a module-level docstring to explain this file's purpose.

File: training/scripts/convert_ds_to_hf.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    parser = argparse.ArgumentParser(description="Convert DeepSpeed Zero shards back to HF formatted checkpoints.")
+    parser.add_argument("--base_model_name_or_path", type=str, required=True,
+                        help="Path or Hub ID of the original model (e.g., meta-llama/Meta-Llama-3-8B)")
+    parser.add_argument("--ds_checkpoint_dir", type=str, required=True,
+                        help="Path to the directory containing the 'global_stepXXX' folder from DeepSpeed (e.g., output/checkpoint-500)")
+    parser.add_argument("--output_dir", type=str, required=True,
+                        help="Target directory to save the clean HF model files.")
+    parser.add_argument("--max_shard_size", type=str, default="5GB",
+                        help="Maximum size of each individual weight shard (e.g., '5GB' or '10GB')")
+    parser.add_argument("--safe_serialization", action="store_true", default=True,
```
Comment (elvisnunez7) [inline training/scripts/convert_ds_to_hf.py]: Is this used?

File: training/scripts/convert_ds_to_hf.py
Comment (elvisnunez7) [inline training/scripts/convert_ds_to_hf.py:1]: Can we rename this file to something like `convert_ds_to_hf.py`?

Comment (elvisnunez7) [inline training/scripts/convert_ds_to_hf.py:1]: Check out `training/src/hmf/launcher.py`, which define some CLI commands that we routinely use in HMF. It would be convenient to be able to do something like `hmf ds2hf --base-model ... --ds-checkpoint ... --output-dir ....` which would call this script so that users don't have to go digging for it.

**Review by elvisnunez7 — CHANGES_REQUESTED** [review]
Thanks for the PR, @dhananjay-ram! Left some comments that need to be addressed before merging.

File: training/src/hmf/train/pt/trainer.py
Comment (elvisnunez7) [inline training/src/hmf/train/pt/trainer.py:1]: `_save` should also be added to our other commonly used trainers, like `sft/trainer.py` and `hkd/trainer.py`.

File: training/docker/patches/async_torch_checkpoint_engine.py (hunk 1 above)
Comment (bbowman223) [reply training/docker/patches/async_torch_checkpoint_engine.py]: For corruption of GPU -> CPU transfer, is that an issue here?  In the script you move weights to CPU then have a torch.cuda.synchronize() barrier.  Then _DS_ASYNC_POOL is called for the RAM --> Disk transfer where the parallelization is happening.  In any case, I am fine to keep as is if it works well.

Comment (dhananjay-ram) [reply training/docker/patches/async_torch_checkpoint_engine.py]: thanks Ben, I tested with ThreadPoolExecutor in the beginning and it was slower.

File: training/docker/patches/async_torch_checkpoint_engine.py
Comment (dhananjay-ram) [reply training/docker/patches/async_torch_checkpoint_engine.py:1]: done

Comment (dhananjay-ram) [reply training/docker/patches/async_torch_checkpoint_engine.py:1]: done

File: training/docker/patches/async_torch_checkpoint_engine.py (hunk 2 above)
Comment (dhananjay-ram) [reply training/docker/patches/async_torch_checkpoint_engine.py]: done

File: training/src/hmf/train/pt/trainer.py (hunk 3 above)
Comment (dhananjay-ram) [reply training/src/hmf/train/pt/trainer.py]: done

File: training/src/hmf/train/pt/trainer.py (hunk 4 above)
Comment (dhananjay-ram) [reply training/src/hmf/train/pt/trainer.py]: done

File: training/src/hmf/train/pt/trainer.py (hunk 7 above)
Comment (dhananjay-ram) [reply training/src/hmf/train/pt/trainer.py]: done

File: training/src/hmf/train/pt/trainer.py
Comment (dhananjay-ram) [reply training/src/hmf/train/pt/trainer.py:1]: done

File: training/scripts/convert_ds_to_hf.py
Comment (dhananjay-ram) [reply training/scripts/convert_ds_to_hf.py:1]: done

File: training/scripts/convert_ds_to_hf.py (hunk 8 above)
Comment (dhananjay-ram) [reply training/scripts/convert_ds_to_hf.py]: you can use it to choose the saving format

**Comment by elvisnunez7** [comment]
Made some changes to complete this PR, summarized below. @bbowman223, PTAL.

### Bug fixes

- **Missing imports in `async_torch_checkpoint_engine.py`**: `os`, `sys`, `signal` were used but never imported. The error handler would crash on checkpoint write failure. Added these imports.
- **Corrupt checkpoints on exit**: Background writes were never awaited, so if training finished before the disk I/O completed, checkpoint files would be truncated and unreadable. Added `atexit` drain to block until all pending writes complete before the process exits.
- **`_save` always skipped HF saves**: The override was unconditional, breaking backward compatibility for users not using async checkpointing (i.e., we would not be able to replicate the saving behavior we had before this PR). Added a gate so that async checkpointing only runs if `async_ckpt` is enabled in the DeepSpeed config.
- **`self.args.max_steps` bug**: When users specify `num_train_epochs` instead of `max_steps`, `self.args.max_steps` is `-1`, causing every checkpoint to trigger a "final save". Changed to `self.state.max_steps` which is always the computed total.
- **Dockerfile COPY path**: Referenced `patches/async_torch_checkpoint_engine.py` but the file lives in `plugins/`.
- **`ds_z3_config_opt_async.json`**: Had `async_ckpt: false` despite the filename suggesting otherwise. Changed to true.
- **DeepSpeed `zero_to_fp32.py` shape bug**: `shape.numel()` crashes when shapes are stored as tuples. Added a patch to normalize shapes across all ZeRO stages (frozen and trainable params).

### New features

- **`hmf ds2hf` CLI command**: Users can now convert DeepSpeed sharded checkpoints to HF format via `hmf ds2hf` without locating the standalone script. Removed the standalone `convert_ds_to_hf.py` script.

### Improvements

- **Rank-guarded prints**: Checkpoint engine and DS patch only log from rank 0 to avoid log spam.
- **Rank in worker messages**: Background IO worker prints include the rank for easier debugging.
- **Removed dead `safe_serialization` arg**: Not used. We always save as safetensors.

### Documentation

- Added "Async Checkpointing" section to `docs/Training.md` covering usage, `hmf ds2hf`, and training resumption.
